### ISSUE TITLE
[CLIFF] 如何在原始图像上旋转GT 2d关键点？

### ISSUE THREAD

**Issue by LastXuanDog** [huawei-noah/noah-research#120]
@zhihaolee  
- 因为在训练的数据处理时会进行数据增强，其中有个旋转的数据增强操作，在原始分辨率下该如何旋转GT 2d keypoints？
- 还有，我注意到您[crop](https://github.com/huawei-noah/noah-research/blob/4853d9f0ef31859c4dcf44f5b56a12798ce5c6b9/CLIFF/common/imutils.py#L52)这里没有像SPIN一样使用rot参数，您是在训练过程中没有使用旋转的数据增强操作么?
- 我使用[uncrop](https://github.com/nkolot/SPIN/blob/5c796852ca7ca7373e104e8489aa5864323fbf84/utils/imutils.py#L82-L106)来把256x192的图片还原到原始size的图片下，但好像位置和尺寸都有点问题。

![image](https://user-images.githubusercontent.com/113993987/192445036-34ee0688-71bf-4d1f-8845-4d30d4e45e39.png)

**Comment by zhihaolee** [comment]
1. 跟之前SPIN的操作一样，以crop中心为旋转中心即可。
2. 我训练时有旋转的数据增强，只是开源代码里没放上这部分代码而已。
3. 需要注意关于尺寸的变量里哪个是宽，哪个是高。

**Comment by LastXuanDog** [comment]
我试了下，确实维度什么的搞反了

**Comment by LastXuanDog** [comment]
@zhihaolee 
![image](https://user-images.githubusercontent.com/113993987/192535058-42e464dd-9147-4907-a133-18c4e6a0ecf4.png)

有些关键点经过旋转后会到画框外，这些点算2d_keypoint_loss的时候需要计算loss么

**Comment by zhihaolee** [comment]
1. 画框外的2D关键点也一样计算loss。
2. 有黑边是正常的。
3. 权重不需要变化。

**Comment by zhihaolee** [comment]
1. 长宽分别除以256和192。
2. 请参考https://github.com/huawei-noah/noah-research/blob/4853d9f0ef31859c4dcf44f5b56a12798ce5c6b9/CLIFF/common/utils.py#L29-L47
3. 请参考EFT的代码。
4. 在正中间，即第100个epoch。

**Comment by LastXuanDog** [comment]
@zhihaolee 
> 如何修改[convert_weak_perspective_to_perspective](https://github.com/mkocabas/PARE/blob/6e0caca86c6ab49ff80014b661350958e5b72fd8/pare/utils/geometry.py#L309)函数可以转换到256x192的透视相机的参数？
对应的256x192cam参数是这样么
'''
bs = 256 * crop_cam[:, 0] + 1e-9
tz = 2 * focal_length / bs
tx = crop_cam[:, 1]
ty = crop_cam[:, 2]
full_cam = torch.stack([tx, ty, tz], dim=-1)
'''

**Comment by zhihaolee** [comment]
是的 @LastXuanDog
